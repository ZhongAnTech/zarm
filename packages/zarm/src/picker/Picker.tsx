import React, { Component } from 'react';
import isEqual from 'lodash/isEqual';
import Popup from '../popup';
import PickerView from '../picker-view';
import BasePickerProps from './PropsType';
import parseProps from '../picker-view/utils/parseProps';
import removeFnFromProps from '../picker-view/utils/removeFnFromProps';
import { DataSource } from '../picker-view/PropsType';
import { WheelValue, WheelItem } from '../wheel/PropsType';

export interface PickerProps extends BasePickerProps {
  prefixCls?: string;
  className?: string;
}

export interface PickerState {
  value: Array<WheelValue>;
  objValue: Array<WheelItem>;
  dataSource: DataSource;
  tempObjValue?: Array<WheelItem>;
  tempValue?: Array<WheelValue>;
  stopScroll?: boolean;
}

export default class Picker extends Component<PickerProps, PickerState> {
  static defaultProps = {
    dataSource: [],
    prefixCls: 'za-picker',
    valueMember: 'value',
    cols: Infinity,
    maskClosable: true,
    itemRender: (data) => data.label,
    destroy: false,
  };

  state: PickerState = { ...parseProps.getSource(this.props), stopScroll: false };

  static getDerivedStateFromProps(props, state) {
    if (
      !isEqual(
        removeFnFromProps(props, ['onOk', 'onCancel', 'onChange']),
        removeFnFromProps(state.prevProps, ['onOk', 'onCancel', 'onChange']),
      )
    ) {
      return {
        prevProps: props,
        ...parseProps.getSource(props),
        tempValue: parseProps.getSource(props).value,
        tempObjValue: parseProps.getSource(props).objValue,
      };
    }

    return null;
  }

  onChange = (selected) => {
    const { valueMember, onChange } = this.props;
    const value = selected.map((item) => item[valueMember!]);
    this.setState({ value, objValue: selected });

    if (typeof onChange === 'function') {
      onChange(selected);
    }
  };

  onCancel = () => {
    const { onCancel } = this.props;
    const { tempValue = [], tempObjValue = [] } = this.state;
    this.setState({
      value: tempValue,
      objValue: tempObjValue,
    });
    if (typeof onCancel === 'function') {
      onCancel();
    }
  };

  onOk = () => {
    const { onOk } = this.props;
    this.setState(
      {
        stopScroll: true,
      },
      () => {
        this.setState(
          {
            stopScroll: false,
          },
          () => {
            if (typeof onOk === 'function') {
              onOk(this.state.objValue);
            }
          },
        );
      },
    );
  };

  render() {
    const {
      prefixCls,
      className,
      cancelText,
      okText,
      title,
      locale,
      maskClosable,
      mountContainer,
      destroy,
      onOk,
      onCancel,
      visible,
      ...others
    } = this.props;
    const { value, stopScroll = false } = this.state;
    const noop = () => {};
    return (
      <Popup
        className={className}
        visible={visible}
        onMaskClick={maskClosable ? this.onCancel : noop}
        mountContainer={mountContainer}
        destroy={destroy}
      >
        <div
          className={prefixCls}
          onClick={(e) => {
            e.stopPropagation();
          }}
        >
          <div className={`${prefixCls}__header`}>
            <div className={`${prefixCls}__cancel`} onClick={this.onCancel}>
              {cancelText || locale!.cancelText}
            </div>
            <div className={`${prefixCls}__title`}>{title || locale!.title}</div>
            <div className={`${prefixCls}__submit`} onClick={this.onOk}>
              {okText || locale!.okText}
            </div>
          </div>
          <PickerView {...others} value={value} onChange={this.onChange} stopScroll={stopScroll} />
        </div>
      </Popup>
    );
  }
}
