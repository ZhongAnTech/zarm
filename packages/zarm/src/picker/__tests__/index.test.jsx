import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Picker from '../index';

describe('Picker', () => {
  const fakeTimers = () => {
    performance.timing = {};
    performance.timing.navigationStart = 0;
  };
  fakeTimers();

  it('Picker render visible', () => {
    const wrapper = mount(
      <Picker
        prefixCls="za-picker"
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        visible
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render custom label', () => {
    const wrapper = mount(
      <Picker
        dataSource={[
          {
            code: '1',
            name: '北京市',
            children: [
              { code: '11', name: '海淀区' },
              { code: '12', name: '西城区' },
            ],
          },
          {
            code: '2',
            name: '上海市',
            children: [
              { code: '21', name: '黄埔区' },
              { code: '22', name: '虹口区' },
            ],
          },
        ]}
        valueMember="code"
        itemRender={(data) => data.name}
      />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should trigger onOk when press ok button', () => {
    jest.useFakeTimers();
    const onOkFn = jest.fn();

    const wrapper = mount(
      <Picker
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
        visible
        value={['1', '12']}
        onOk={onOkFn}
      />,
    );

    wrapper.find('.za-picker__submit').simulate('click');
    jest.runAllTimers();
    expect(onOkFn).toBeCalled();
  });

  it('should trigger onCancel when press cancel button', () => {
    const onCancelFn = jest.fn();

    const wrapper = mount(
      <Picker
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
        visible
        defaultValue={['1', '12']}
        onCancel={onCancelFn}
      />,
    );

    wrapper.find('.za-picker__cancel').simulate('click');
    expect(onCancelFn).toBeCalled();
  });

  // it('should trigger onMaskClick when click mask', () => {
  //   const onOkFn = jest.fn();
  //   const onMaskClick = jest.fn();

  //   const wrapper = mount(
  //     <Picker
  //       dataSource={[
  //         {
  //           value: '1',
  //           label: '选项一',
  //           children: [
  //             { value: '11', label: '选项一' },
  //             { value: '12', label: '选项二' },
  //           ],
  //         },
  //         {
  //           value: '2',
  //           label: '选项一',
  //           children: [
  //             { value: '21', label: '选项一' },
  //             { value: '22', label: '选项二' },
  //           ],
  //         },
  //       ]}
  //       visible
  //       defaultValue={['1', '12']}
  //       onOk={onOkFn}
  //       onMaskClick={onMaskClick}
  //     />, { attachTo: window.domNode }
  //   );

  //   wrapper.find('.za-mask').simulate('click');
  //   expect(onMaskClick).toBeCalled();
  // });

  it('receive new dataSource', () => {
    const wrapper = shallow(
      <Picker
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    wrapper.setProps({
      dataSource: [
        { value: 'a', label: '选项一' },
        { value: 'b', label: '选项二' },
        { value: 'c', label: '选项三' },
      ],
    });
  });

  it('receive new value', () => {
    const wrapper = shallow(
      <Picker
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    wrapper.setProps({ value: '1' });
  });

  it('receive new cascader dataSource', () => {
    const wrapper = shallow(
      <Picker
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
      />,
    );

    wrapper.setProps({
      dataSource: [
        {
          value: '3',
          label: '选项一',
          children: [
            { value: '31', label: '选项一' },
            { value: '32', label: '选项二' },
          ],
        },
        {
          value: '4',
          label: '选项一',
          children: [
            { value: '41', label: '选项一' },
            { value: '42', label: '选项二' },
          ],
        },
      ],
    });
  });

  // it('StackPicker', () => {
  //   jest.useFakeTimers();
  //   const wrapper = mount(
  //     <Picker.Stack
  //       dataSource={District}
  //       />
  //   );
  //   expect(toJson(wrapper)).toMatchSnapshot();
  //   wrapper.setProps({ value: ['安徽省', '安庆市', '大观区'] });
  //   jest.runAllTimers();
  //   wrapper.unmount();
  // });

  // it('StackPicker init value', () => {
  //   const wrapper = mount(
  //     <Picker.Stack
  //       dataSource={District}
  //       value={['340000', '340800', '340803']}
  //       />
  //   );
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });

  // it('StackPicker changeValue', () => {
  //   const onChangeFn = jest.fn();
  //   const wrapper = mount(
  //     <Picker.Stack
  //       dataSource={District}
  //       onOk={onChangeFn}
  //       />
  //   );

  //   wrapper.find('.za-picker-stack-column').at(0).simulate('click');
  //   wrapper.find('.za-picker-stack-item').at(0).simulate('click');
  //   expect(onChangeFn).toBeCalled();
  // });

  // it('StackPicker maskClick', () => {
  //   const wrapper = mount(
  //     <Picker.Stack
  //       dataSource={District}
  //       />
  //   );
  //   wrapper.find('.za-picker-input').simulate('click');
  //   wrapper.find('.za-picker-cancel').simulate('click');
  //   wrapper.find('.za-mask').simulate('click');
  // });
});
