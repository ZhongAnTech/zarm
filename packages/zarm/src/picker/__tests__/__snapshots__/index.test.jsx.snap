// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker Picker render visible 1`] = `
<ForwardRef(forwardRef)
  dataSource={
    Array [
      Object {
        "label": "选项一",
        "value": "1",
      },
      Object {
        "label": "选项二",
        "value": "2",
      },
    ]
  }
  prefixCls="za-picker"
  visible={true}
>
  <ConfigReceiver
    dataSource={
      Array [
        Object {
          "label": "选项一",
          "value": "1",
        },
        Object {
          "label": "选项二",
          "value": "2",
        },
      ]
    }
    forwardedRef={null}
    prefixCls="za-picker"
    visible={true}
  >
    <Picker
      cols={Infinity}
      dataSource={
        Array [
          Object {
            "label": "选项一",
            "value": "1",
          },
          Object {
            "label": "选项二",
            "value": "2",
          },
        ]
      }
      destroy={false}
      itemRender={[Function]}
      locale={
        Object {
          "cancelText": "取消",
          "okText": "确定",
          "title": "请选择",
        }
      }
      localeCode="zh-CN"
      maskClosable={true}
      prefixCls="za-picker"
      safeIphoneX={false}
      valueMember="value"
      visible={true}
    >
      <Popup
        destroy={false}
        lockScroll={true}
        onMaskClick={[Function]}
        visible={true}
      >
        <Portal
          animationDuration={200}
          animationType="fade"
          direction="bottom"
          handlePortalUnmount={[Function]}
          mask={true}
          onMaskClick={[Function]}
          prefixCls="za-popup"
          visible={true}
        >
          <Trigger
            disabled={false}
            onClose={[Function]}
            visible={true}
          >
            <Portal
              containerInfo={
                <div
                  class="za-popup-container"
                >
                  <div
                    class="za-mask za-fade-enter za-mask--normal"
                    style="animation-duration: 200ms;"
                  />
                  <div
                    class="za-popup__wrapper"
                    data-width="0"
                    role="dialog"
                  >
                    <div
                      class="za-popup za-popup--bottom za-popup--show"
                      data-width="0"
                      role="document"
                      style="transition-duration: 200ms; transition-property: transform;"
                    >
                      <div
                        class="za-picker"
                      >
                        <div
                          class="za-picker__header"
                        >
                          <div
                            class="za-picker__cancel"
                          >
                            取消
                          </div>
                          <div
                            class="za-picker__title"
                          >
                            请选择
                          </div>
                          <div
                            class="za-picker__submit"
                          >
                            确定
                          </div>
                        </div>
                        <div
                          class="za-picker-view"
                        >
                          <div
                            class="za-picker-view__content"
                          >
                            <div
                              class="za-wheel"
                            >
                              <div
                                class="za-wheel__content"
                                style="transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-property: transform; transition-duration: 800ms; transform: translateX(0px) translateY(NaNpx);"
                              >
                                <div
                                  class="za-wheel__item za-wheel__item--selected"
                                  style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
                                >
                                  选项一
                                </div>
                                <div
                                  class="za-wheel__item"
                                  style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
                                >
                                  选项二
                                </div>
                              </div>
                            </div>
                          </div>
                          <div
                            class="za-picker-view__mask za-picker-view__mask--top"
                          />
                          <div
                            class="za-picker-view__mask za-picker-view__mask--bottom"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              }
            >
              <Mask
                className="za-fade-enter"
                style={
                  Object {
                    "WebkitAnimationDuration": "200ms",
                    "animationDuration": "200ms",
                  }
                }
                type="normal"
                visible={true}
              >
                <div
                  className="za-mask za-fade-enter za-mask--normal"
                  style={
                    Object {
                      "WebkitAnimationDuration": "200ms",
                      "animationDuration": "200ms",
                    }
                  }
                />
              </Mask>
              <div
                className="za-popup__wrapper"
                data-width={0}
                onClick={[Function]}
                role="dialog"
                style={Object {}}
              >
                <div
                  className="za-popup za-popup--bottom"
                  data-width={0}
                  role="document"
                  style={
                    Object {
                      "WebkitTransitionDuration": "200ms",
                      "WebkitTransitionProperty": "transform",
                      "transitionDuration": "200ms",
                      "transitionProperty": "transform",
                    }
                  }
                >
                  <div
                    className="za-picker"
                    onClick={[Function]}
                  >
                    <div
                      className="za-picker__header"
                    >
                      <div
                        className="za-picker__cancel"
                        onClick={[Function]}
                      >
                        取消
                      </div>
                      <div
                        className="za-picker__title"
                      >
                        请选择
                      </div>
                      <div
                        className="za-picker__submit"
                        onClick={[Function]}
                      >
                        确定
                      </div>
                    </div>
                    <PickerView
                      cols={Infinity}
                      dataSource={
                        Array [
                          Object {
                            "label": "选项一",
                            "value": "1",
                          },
                          Object {
                            "label": "选项二",
                            "value": "2",
                          },
                        ]
                      }
                      disabled={false}
                      itemRender={[Function]}
                      localeCode="zh-CN"
                      onChange={[Function]}
                      prefixCls="za-picker-view"
                      safeIphoneX={false}
                      stopScroll={false}
                      value={
                        Array [
                          "1",
                        ]
                      }
                      valueMember="value"
                    >
                      <div
                        className="za-picker-view"
                      >
                        <div
                          className="za-picker-view__content"
                        >
                          <Wheel
                            dataSource={
                              Array [
                                Object {
                                  "label": "选项一",
                                  "value": "1",
                                },
                                Object {
                                  "label": "选项二",
                                  "value": "2",
                                },
                              ]
                            }
                            disabled={false}
                            itemRender={[Function]}
                            key="0"
                            onChange={[Function]}
                            stopScroll={false}
                            value="1"
                            valueMember="value"
                          >
                            <div
                              className="za-wheel"
                            >
                              <div
                                className="za-wheel__content"
                              >
                                <div
                                  className="za-wheel__item za-wheel__item--selected"
                                  key="0"
                                >
                                  选项一
                                </div>
                                <div
                                  className="za-wheel__item"
                                  key="1"
                                >
                                  选项二
                                </div>
                              </div>
                            </div>
                          </Wheel>
                        </div>
                        <div
                          className="za-picker-view__mask za-picker-view__mask--top"
                        />
                        <div
                          className="za-picker-view__mask za-picker-view__mask--bottom"
                        />
                      </div>
                    </PickerView>
                  </div>
                </div>
              </div>
            </Portal>
          </Trigger>
        </Portal>
      </Popup>
    </Picker>
  </ConfigReceiver>
</ForwardRef(forwardRef)>
`;

exports[`Picker render custom label 1`] = `
<ForwardRef(forwardRef)
  dataSource={
    Array [
      Object {
        "children": Array [
          Object {
            "code": "11",
            "name": "海淀区",
          },
          Object {
            "code": "12",
            "name": "西城区",
          },
        ],
        "code": "1",
        "name": "北京市",
      },
      Object {
        "children": Array [
          Object {
            "code": "21",
            "name": "黄埔区",
          },
          Object {
            "code": "22",
            "name": "虹口区",
          },
        ],
        "code": "2",
        "name": "上海市",
      },
    ]
  }
  itemRender={[Function]}
  valueMember="code"
>
  <ConfigReceiver
    dataSource={
      Array [
        Object {
          "children": Array [
            Object {
              "code": "11",
              "name": "海淀区",
            },
            Object {
              "code": "12",
              "name": "西城区",
            },
          ],
          "code": "1",
          "name": "北京市",
        },
        Object {
          "children": Array [
            Object {
              "code": "21",
              "name": "黄埔区",
            },
            Object {
              "code": "22",
              "name": "虹口区",
            },
          ],
          "code": "2",
          "name": "上海市",
        },
      ]
    }
    forwardedRef={null}
    itemRender={[Function]}
    valueMember="code"
  >
    <Picker
      cols={Infinity}
      dataSource={
        Array [
          Object {
            "children": Array [
              Object {
                "code": "11",
                "name": "海淀区",
              },
              Object {
                "code": "12",
                "name": "西城区",
              },
            ],
            "code": "1",
            "name": "北京市",
          },
          Object {
            "children": Array [
              Object {
                "code": "21",
                "name": "黄埔区",
              },
              Object {
                "code": "22",
                "name": "虹口区",
              },
            ],
            "code": "2",
            "name": "上海市",
          },
        ]
      }
      destroy={false}
      itemRender={[Function]}
      locale={
        Object {
          "cancelText": "取消",
          "okText": "确定",
          "title": "请选择",
        }
      }
      localeCode="zh-CN"
      maskClosable={true}
      prefixCls="za-picker"
      safeIphoneX={false}
      valueMember="code"
    >
      <Popup
        destroy={false}
        lockScroll={true}
        onMaskClick={[Function]}
        visible={false}
      />
    </Picker>
  </ConfigReceiver>
</ForwardRef(forwardRef)>
`;
